File: 96-unique-binary-search-trees/96-unique-binary-search-trees.py
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [0]*(n+1)
        dp[0] = 1
        for i in range(1,n+1):
            if i == 1:
                dp[i] = 1
            else:
                for j in range(1,i+1):
                    dp[i] += dp[j-1] * dp[i-j]
        return dp[n]
        

File: 96-unique-binary-search-trees/96-unique-binary-search-trees.java
class Solution {
    Map<Pair, Integer> memo = new HashMap<>();
    public int numTrees(int n) {
        return numSubTrees(1, n);
    }
    
    public int numSubTrees(int start, int end){
        if(start > end)return 1;
        int res = 0;
        Pair key = new Pair(start, end);
        if(memo.containsKey(key))return memo.get(key);
        
        for(int i = start; i < end + 1; i++){
            int leftNum = numSubTrees(start, i - 1);
            int rightNum = numSubTrees(i + 1, end);
            res += (leftNum * rightNum);
        }
        memo.put(key, res);
        return res;
    }
}